receivers-in-go.md

```diff
@@ -0,0 +1,257 @@
+
+```go
+type Server struct {
+    ...
+}
+
+func (srv *Server) Close() error {
+    ...
+}
+
+func (srv Server) Name() string {
+    ...
+}
+```
+
+Every time I start writing somekind of server, it starts out looking like this.  
+It's a type that holds information about our server, like a `net.Listener` 
+socket, and it has a `Close()` method that shuts down the server. Easy enough.
+
+The *receiver* of the `Close()` method is the `(srv *Server)` part. This says 
+that inside of the `Close()` method declaration, the scope will have an `srv` 
+variable that is a reference to the instance of the `Server` that it's being 
+called on.  That is:
+
+```go
+myServer := &Server{}
+myServer.Close()
+```
+
+In this case, the `s` that is referenced inside of the `myServer.Close()` is 
+effectively the same variable as `myServer`. They're both references to the same 
+`Server` instance.
+
+
+## Naming the receiver
+
+In a lot of Go code, it's common to use the first letter or a short abbreviation 
+as the name of the receiver. If the name of the struct is `Server`, we'll 
+usually see `s` or `srv` or even `server`. All of these are fine.
+
+But why not `self` or `this`? Coming from Python, or Ruby, or JavaScript, it's 
+tempting to do something like:
+
+```go
+func (this *Server) Close() error {
+    ...
+}
+```
+
+That's one less decision to make every time we declare a struct. *All* of our 
+methods could use the same receiver. Any time we see `this` in the code, we'll 
+*know* that we're talking about the receiver, not some random local variable. It 
+will be GREAT!.. or will it?
+
+
+## Facts about methods and receivers
+
+While we can call a method on a type instance and get the receiver implicitly, 
+it can also be called explicitly:
+
+```go
+myServer := &Server{}
+Server.Name(myServer)
+(*Server).Close(&myServer)
+```
+
+These functions can be passed around as references just like any other function, 
+with the implicit receiver or without:
+
+```go
+withReceiver := myServer.Name
+without := Server.Name
+```
+
+Receivers can be passed by reference or passed by value.
+
+```go
+func (byValue Server) Hello() { ... }
+func (byReference *Server) Bye() { ... }
+```
+
+This is to illustrate that struct methods in Go are merely thin sugar over 
+traditional C-style struct helper declarations, an equivalent C method might 
+look like this:
+
+```c
+void server_close(server *srv) { ... }
+```
+
+Go helps by namespacing the methods and implicitly passing the receiver when 
+called on an instance, but otherwise there is very little magic going on.
+
+In other languages where `this` and `self` is a thing (Python, Ruby, JavaScript, 
+and so on) it's a much more complicated situation. These are not vanilla local 
+variables wearing fancy pants. The thing we might expect `this` to represent 
+inside of a method could actually represent something very different once 
+inheretance or metaclasses had their way. In effect, it might not make any sense 
```

Comment:
Wait, how does it follow that therefore you should give contextual names to self? I feel like I'm missing a premise here.